### ISSUE TITLE
Enhancement - Confirm the recovery of the password via E-Mail

### ISSUE THREAD

**Issue by ghost** [Xephi/AuthMeReloaded#472]
Hello,

I have found a pretty important enchantment for your plugin, you should add the confirmation of the password recovery via E-Mail, because now a player with an offline-version of Minecraft can just connect to the server with the nick of a player that they don't like, recover the password via E-Mail (assuming they know the E-Mail address of their "victim"), and so the players can annoy other players that they don't like, because the victim can not log in to his account before he checks his E-Mail and see that he's password got recovered.

Thank you.

**Comment by RoinujNosde** [comment]
What if the password is not changed on a recovery?
For instance:
Old password: 12345
New password: 54321

If I login with 12345 means that I was able to remember my password, so the new one is not necessary.
The old one would only be dumped if the new is used.

**Comment by RoinujNosde** [comment]
Other thing that can prevent this is secret answers.
In order to request a new password, player would have to answer correctly to questions he created.

**Comment by Naxter** [comment]
Hmm, I think the first option were pretty good, setting secret answers would in my opinion take too much time for the players. But I think you should implement some spam-security system, so that a player couldn't just spam the E-Mail address with new passwords of the real account owner.

**Comment by ljacqu** [comment]
Assigning to myself to come up with some technical solutions within the next two days

**Comment by ljacqu** [comment]
Request for feedback—@Naxter @RoinujNosde @sgdc3

My suggestion: sending a code in the email that the user needs to supply to be able to change password.
--> User does /email recover --> gets email with "hi %player, use /recover 123456 [pass] [pass] to reset your email", where 123456 = some random, sufficiently long code!

In database this only requires two additional fields: recoveryCode + timestamp because we need to limit this to, say, 3-7 days. We can even hash the code, say, sha256(nickname + code) in the db for additional security.

-

Secret answers are a good idea but I would strongly advise not to do that: as @Naxter says it takes effort for the user to set up; users will probably enter obvious info just to deal with it; we need to be able to deal with the input if it's multiple words, which we currently can't do from a command handling perspective; having secret answers in the DB means there is _always_ a backdoor for a password reset; we need logic for adding/managing those answers.

**Comment by RoinujNosde** [comment]
@ljacqu Recovery code is a great idea :+1:

**Comment by sgdc3** [comment]
@ljacqu Yeah, it seems enough user friendly ;)

**Comment by ljacqu** [comment]
Hello @sgdc3 @RoinujNosde @Platinteufel and @Twonox, allow me to @-mention you since the original poster of this issue is now a ghost ;)

I've implemented a confirmation code for password recovery by email. Now the user first gets a code by mail that he needs to enter, and _then_ his password is changed an another mail is sent. Can be turned off by setting the code length setting to 0.
This took longer to implement than expected, and there's quite some room for improvement. I'd be very thankful for some feedback and suggestions.

To reset password, this happens now:
Player: /email recover [email]
Server: A recovery code was sent to your email
Player: /email recover [email] [code]
Server: Your password has been changed and sent by mail
